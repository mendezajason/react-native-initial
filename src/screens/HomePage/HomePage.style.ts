import {StyleSheet, ViewStyle, TextStyle} from 'react-native';

interface Style {
  container: ViewStyle;
  item: ViewStyle;
  name: TextStyle;
  email: TextStyle;
}

export default StyleSheet.create<Style>({
  container: {
    flex: 1,
    padding: 10,
  },
  item: {
    paddingHorizontal: 15,
    paddingVertical: 7,
  },
  name: {
    color: '#282c44',
    fontSize: 19,
    fontWeight: '700',
  },
  email: {
    color: '#444444',
    fontSize: 15,
  },
});
