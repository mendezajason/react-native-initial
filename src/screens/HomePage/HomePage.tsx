import React, {type PropsWithChildren, useState, useEffect} from 'react';
import {
  View,
  Text,
  ScrollView,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import {useDispatch} from 'react-redux';
import {setUserData} from '../../features/users/selectUser';
import styles from './HomePage.style';

export type UserType = {
  id: number;
  name: string;
  username: string;
  email: string;
  address: {
    street: string;
    suite: string;
    city: string;
    zipcode: string;
    geo: {
      lat: string;
      lng: string;
    };
  };
  phone: string;
  website: string;
  company: {
    name: string;
    catchPhrase: string;
    bs: string;
  };
};
const HomePage: React.FC<PropsWithChildren<{navigation: any}>> = ({
  navigation,
}) => {
  const [users, setUsers] = useState<UserType[]>([]);
  const [loading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    fetchUsers();
  }, []);

  const fetchUsers = async () => {
    setLoading(false);
    const url = 'https://jsonplaceholder.typicode.com/users';
    const requestOptions = {
      method: 'GET',
      redirect: 'follow',
    };
    try {
      const response = await fetch(url, requestOptions);
      const json = await response.json();
      setUsers(json);
    } catch (error) {
      console.error(error);
    } finally {
      setLoading(true);
    }
  };

  return (
    <View style={[styles.container]}>
      <ScrollView>
        {loading &&
          users?.length > 0 &&
          users?.map((value: UserType, index: number) => (
            <ItemList
              key={index}
              id={value?.id || 0}
              navigation={navigation}
              name={value?.name || '----'}
              email={value?.email || '----'}
            />
          ))}
        {!loading && <ActivityIndicator />}
      </ScrollView>
    </View>
  );
};

export default HomePage;

const ItemList: React.FC<
  PropsWithChildren<{id: number; name: string; email: string; navigation: any}>
> = ({id, name, email, navigation}) => {
  const dispatch = useDispatch();

  const onPress = async () => {
    await fetchUser();
    navigation.navigate('Detail');
  };

  const fetchUser = async () => {
    const url = `https://jsonplaceholder.typicode.com/users/${id}`;
    const requestOptions = {
      method: 'GET',
      redirect: 'follow',
    };
    try {
      const response = await fetch(url, requestOptions);
      const json = await response.json();
      dispatch(setUserData(json));
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <TouchableOpacity onPress={() => onPress()}>
      <View style={[styles.item]}>
        <Text style={[styles.name]}>{name}</Text>
        <Text style={[styles.email]}>{email}</Text>
      </View>
    </TouchableOpacity>
  );
};
