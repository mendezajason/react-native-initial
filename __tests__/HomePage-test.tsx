/**
 * @format
 */

import 'react-native';
import React from 'react';
import HomePage from '../src/screens/HomePage/HomePage';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const home = renderer.create(<HomePage />).toJSON();
  expect(home).toMatchSnapshot();
});

