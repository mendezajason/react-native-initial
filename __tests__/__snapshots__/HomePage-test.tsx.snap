// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "padding": 10,
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <ActivityIndicator />
    </View>
  </RCTScrollView>
</View>
`;
